03week/towersOfHanoi.js

```diff
@@ -17,28 +23,85 @@ function printStacks() {
   console.log("a: " + stacks.a);
   console.log("b: " + stacks.b);
   console.log("c: " + stacks.c);
+  // console.log("and so it begins....");
+  // console.log(startStack[startStack.length-1]);
 }
 
-function movePiece() {
-  // Your code here
 
+// ========================================================
+// ========= Checkpoint1 Assignment Starts Here ===========
+// ========================================================
+
+
+// ========== Function movePiece =========
+function movePiece(startStack, endStack) {
+  // move the top piece from startStack to top of endStack
+  let movingPiece = stacks[startStack].pop();
+  stacks[endStack].push(movingPiece);
+  // console.log("moved piece and updated the board");
 }
 
-function isLegal() {
-  // Your code here
 
+// ========= Function isLegal =========
+function isLegal(startStack, endStack) {
+  // if the move is legal return true, a legal move is
+  // when at least 1 piece resides in the startStack and..
+  // when the endStack is empty or...
+  // when the top piece in the start stack is smaller than the top piece in the endstack
+  // else the move is not legal return false
+  let topStartPiece = stacks[startStack][stacks[startStack].length - 1];
+  let topEndPiece = stacks[endStack][stacks[endStack].length - 1];
+  console.log("top start piece:", topStartPiece);
+  console.log("top end piece:", topEndPiece);
+
+  if (stacks[startStack].length > 0 && (topStartPiece < topEndPiece || (topEndPiece === undefined))) {
+    console.log ("the move is legal");
+    movePiece(startStack, endStack);
+    checkForWin();
+    return true;
+  } else {
+    console.log ("the move is not legal");
+    return false;
+  }
 }
 
-function checkForWin() {
-  // Your code here
 
+// ========== Function checkForWin ============
+function checkForWin() {
+  // if the player won, return true
+  // if the player did not win, return false
+  // a win is when all 4 blocks are in stack b or stack c
+  if ((stacks.b.length || stacks.c.length) >=4) {
```

Comment:
that's not really the way to check if either stack have a length of 4
```
if(stacks.b.length >=4 || stacks.c.length >=4) { 
```

Reply:
duly noted, good tip